src/lib/assert.hpp

```diff
@@ -32,6 +32,9 @@ class assertion_error : public std::runtime_error
        : throw assertion_error(#expr, __FILE__, __LINE__, __PRETTY_FUNCTION__))
 #define RUN_DEBUG(expr) void(expr)
 #else
-#define ASSERT(expr) (static_cast<void>(0))
-#define RUN_DEBUG(expr) void(0)
+// type-check the assert expressions even in release mode, but don't run them
+//  (compiler should optimize this away into a no-op)
+// the inner (void)(expr) suppresses the "expression result unused" warning
```

Comment:
Isn't it better to use `if constexpr`? We need C++17 for other things, so it should be possible to use `if constexpr` as well and avoid this preprocessor trickery. Or am I missing something?

Reply:
That should be possible, I just don't know modern C++ that well. :)